### PR TITLE
Dispatch Managing Actor

### PR DESCRIPTION
**Pull request by Marenz**
- **Reset release notes**
- **Add dispatch runner**

### CODE REVIEW

File: tests/test_runner.py
Comment (daniel-zullo-frequenz) [inline tests/test_runner.py]: These lines might no longer be needed

File: tests/test_runner.py
Comment (daniel-zullo-frequenz) [inline tests/test_runner.py]: Probably not needed

File: src/frequenz/dispatch/_actor_runner.py
Comment (daniel-zullo-frequenz) [inline src/frequenz/dispatch/_actor_runner.py]: The function returns None :thinking:

**Review by daniel-zullo-frequenz — COMMENTED** [review]
Just a few really minor comments to check for, LGTM otherwise

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [inline src/frequenz/dispatch/_actor_runner.py]: I'm still not 100% sold on this, I still feel that it might be better to send an actual `Dispatch` object, possibly a merged one (i.e. if we are merging all dispatches with different IDs but the same type, for example, and we have id=1 start=10:00 end=11:00 and id=2 start=11:00 end=None, that would be merged into a dispatch with start=10:00 and end=None).

But since that would be harder to implement I'm OK with starting with this simple approach and see if we need something more complicated in the future.

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [inline src/frequenz/dispatch/_actor_runner.py]: I used this scheme until I realized that is exactly the same as `case unsupported:`.

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [inline src/frequenz/dispatch/_actor_runner.py]: I know this is a separate repo, but I was confused for a second with the config manager. I would try to make a more explicit separation, maybe call this `dispatch_reconfiguration_channel` (or `dispatch_updates_channel`). I actually think `update` is a better name here, this is not about configuration, but about receiving updates to a running dispatch.

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [inline src/frequenz/dispatch/_actor_runner.py]: I don't get what this `supporting_actor` is.

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [inline src/frequenz/dispatch/_actor_runner.py]: I would call this `dispatch_runner = DispatchRunnerActor(...)` for consistency.

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [inline src/frequenz/dispatch/_actor_runner.py]: Do we have usecases for passing multiple actors to this actor?

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [inline src/frequenz/dispatch/_actor_runner.py]: I would use the more idiomatic:

```py
        async for dispatch in self._dispatch_rx:
            await self._handle_dispatch(dispatch=dispatch)
```

And then add a log when a dispatch update is received (like "dispatch update received, processing...", probably with debug level) and finished processing (like "dispatch update processed"). You lose the explicit "waiting" info message, but it is also implicit that it is waiting once it finished processing a dispatch. You also get an automatic info log when the actor is started (via the `Actor` class).

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [inline src/frequenz/dispatch/_actor_runner.py]: Maybe "Stopped by a dispatch".

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [inline src/frequenz/dispatch/_actor_runner.py]: I would call this `DispatchUpdate`. I think calling this configuration is a bit confusing because we use *configuration* for something else, and also I think even should be "created", "updated", etc. For me this class only holds a *dispatch update*.

**Review by llucax — COMMENTED** [review]
I would say all comments are mostly bike shedding, but in particular I think not using *configuration* and use *dispatch update* is important, I think it can get a bit confusing for people dealing with dispatches and configuration (as in config manager) too. I really think we should replace all `*configuration* mentions to *update*/*dispatch update*.

Also I found the example a bit confusing, which ideally it shouldn't be the case, the example should help make things less confusing :P

File: src/frequenz/dispatch/_actor_runner.py
Comment (Marenz) [reply src/frequenz/dispatch/_actor_runner.py]: Dunno, @daniel-zullo-frequenz seems to have required it, that part is from him :D

**Comment by Marenz** [comment]
I don't like `update` much in this context either though. How about `DispatchParameterUpdate`

File: src/frequenz/dispatch/_actor_runner.py
Comment (daniel-zullo-frequenz) [reply src/frequenz/dispatch/_actor_runner.py]: Yes, there is a case where we start/stop more than 2 actors. Is there any reason why you'd like to handle only a single actor?

Comment (llucax) [reply src/frequenz/dispatch/_actor_runner.py]: It feels like it is more natural to map one actor to handling one dispatch, but I'm not strongly opposed. I would like to hear about the details on your use case thought, maybe in private to avoid spamming the PR too much?

**Comment by llucax** [comment]
> I don't like `update` much in this context either though. How about `DispatchParameterUpdate`

Why don't you like `update`? Adding `Parameter` doesn't make it better IMHO, there is no concept of a dispatch parameter in the API. Another option would be `change` instead of `update`, but for me `update` sounds more accurate in this context, this is why I wonder what you don't like about `update`.

File: RELEASE_NOTES.md
Comment (llucax) [inline RELEASE_NOTES.md]: Given this description, `DispatchManagingActor` doesn't sound like a bad option either. But again, bike shedding, I think it is only marginally better.

**Review by llucax — COMMENTED** [review]
I see there was not much progress or replies on my comments. I guess you are focusing on getting shit to work, which sounds totally reasonable, but just saying that I didn't review much in this round.

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [inline src/frequenz/dispatch/_actor_runner.py]: bikeshedding: I'm more and more convinced this is more a `DispatchManagingActor`.

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [inline src/frequenz/dispatch/_actor_runner.py]: bikeshedding: Payload sounds weird to me. For a whole dispatch it makes sense (or at least it was the best we could find) but in this context maybe something like `options` is more intuitive? To be honest I think `options` might be a good choice for the specs too.

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [inline src/frequenz/dispatch/_actor_runner.py]: Following my quest to remove "config" from dispatch...

```suggestion
    """Event emitted when the dispatch changes."""
```

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [inline src/frequenz/dispatch/_actor_runner.py]:
```suggestion
        def __init__(self, updates_channel: Receiver[DispatchUpdate]):
            super().__init__()
            self._updates_channel = updates_channel
            self._dry_run: bool
            self._payload: dict[str, Any]

        async def _run(self) -> None:
            while True:
                update = await self._updates_channel.receive()
                print("Received update:", update)

                self.set_components(update.components)
                self._dry_run = update.dry_run
                self._payload = update.payload
```

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [inline src/frequenz/dispatch/_actor_runner.py]: And again:

```py
        async for dispatch in self._dispatch_rx:
            await self._handle_dispatch(dispatch=dispatch)
```

But actually this example doesn't look very realistic, I guess the actor should do something else than just *listening* to updates. Maybe it is worth adding some other logic to represent that, maybe implement a `ClockActor` that ticks every X time, where X is a dispatch option? So we have a `select()` instead with a `timer` (printing the time) and the `updates_channel`.

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [inline src/frequenz/dispatch/_actor_runner.py]:
```suggestion
        # Create updates channel to receive update events pre-start and mid-run
```

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [inline src/frequenz/dispatch/_actor_runner.py]:
```suggestion
        # Start actor and supporting actor, give each a updates channel receiver
```

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [inline src/frequenz/dispatch/_actor_runner.py]:
```suggestion
            updates_sender=dispatch_updates_channel.new_sender(),
```

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [inline src/frequenz/dispatch/_actor_runner.py]:
```suggestion
        updates_sender: Sender[DispatchUpdate] | None = None,
```

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [inline src/frequenz/dispatch/_actor_runner.py]:
```suggestion
            updates_sender: The sender for dispatch updates.
```

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [inline src/frequenz/dispatch/_actor_runner.py]:
```suggestion
        self._updates_sender = updates_sender
```

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [inline src/frequenz/dispatch/_actor_runner.py]:
```suggestion
                if self._updates_sender is not None:
                    _logger.info("Updated by dispatch %s", dispatch.id)
```

File: src/frequenz/dispatch/_dispatch.py
Comment (llucax) [inline src/frequenz/dispatch/_dispatch.py:195]: For a different PR (or to convert to an issue in the client repo):

At the python client level we could actually very easily use special values to represent `INFINITE` and `INSTANTANEOUS`, so the code gets more readable (without the need for comments).

We could set the type of `duration: timedelta | Duration` where duration is a enum for example, then we can also use `match` like this:

```py
match dispatch.duration:
    case Duration.INFINITE: ...
    case Duration.INSTANTANEOUS: ...
    case timedelta(): ... (here we know it is > 0)
    case unexpected: assert_never(unexpected)
```

Taking it even a bit further, we could have a `PositiveTimedelta` object that even guarantee the time is positive.

File: src/frequenz/dispatch/actor.py
Comment (llucax) [inline src/frequenz/dispatch/actor.py:112]: Also for a separate issue: if this function is used externally, then it should be public. If we want to restrict access only to certain users, one trick I've been using is to use either a `Protocol` for the public interface (where you don't expose the `set_deleted()` method) or just plain inheritance (`Dispatch` without `set_delete()` and a `DeletableDispatch` that inherits from it; internally you use `DeletableDispatch` but when returning it to the users, you return a plain `Dispatch`).

This actually made me think why are are storing state in a immutable object? That looks weird.
Another approach, since the `Dispatch`'s `deleted` `bool` and `running_state_change_synced` are only attached as mutable properties for internal **state** tracking, it might make more sense to separate the **state** from the immutable dispatch, and have a `DispatchState` **mutable** class and a separate `dict` to map dispatch ID to its state.

You can pass the state object to the `Dispatch` constructor and store it as a private property and expose *read-only* properties to access the state, if it needs to be accessed externally, and store the `dict` and mutate it from this actor only.

File: tests/test_frequenz_dispatch.py
Comment (llucax) [inline tests/test_frequenz_dispatch.py]: :warning:  This shouldn't be necessary and I really think we shouldn't go this way. Things should work if every test runs in its own loop and if it doesn't it means some cleanup is missing or messed up. :warning: 

Oh, I just noticed it also goes back from the correct `event_loop_policy()` to the old (and deprecated) `event_loop()` which is supposed to stop working soon :tm:. This is definitely not an option.

**Review by llucax — COMMENTED** [review]
Very nice to see the new scheduling algorithm! :tada: Way better than what we had before.

Lots of comments, so here is a summary:

Things I think needs to be addressed before merging:

1. These commits need a commit message with a body that describes what's going on, it is really hard to review them without more information of what was wrong and how it was fixed:

    * `Fix wrong running behaviors with duration=None`
    * `Rewrite internal architecture to be more flexible`
    * `Add dispatch runner` (less important but it would be nice, is not a trivial class)
  
2. Finish the renaming of `config` -> `update`
3. The hack in the tests (session-wide `event_loop` fixture), we can't really merge this if something is so fucked up that makes tests hang or make the interpreter segfault :grimacing: 
4. Figure out if we really need to support multiple actors, it is still make noise in my head, my feeling is we should encourage having one actor to handle one dispatch type, if that actor needs other supporting actors, it should be the actor itself in charge of spawning them.
5. If we are doing more renaming (like `DispatchManagingActor` and `payload` -> `options`), it might be good to do them now too to avoid having breaking changes in the future.

The rest is minor or too big for this PR, so I will create follow-up issues.

File: tests/test_frequenz_dispatch.py
Comment (Marenz) [reply tests/test_frequenz_dispatch.py]: I am happy to take any working alternative ;)

**Comment by Marenz** [comment]
And I still think "update" sucks and "configuration" was better. "Update" doesn't fit at all in the whole naming scheme

File: src/frequenz/dispatch/_actor_runner.py
Comment (Marenz) [reply src/frequenz/dispatch/_actor_runner.py]: how about `inputs`?

**Comment by Marenz** [comment]
Addressed points 1 - 5

File: src/frequenz/dispatch/_managing_actor.py
Comment (llucax) [inline src/frequenz/dispatch/_managing_actor.py]: I just realized this is tricky. The actor needs to make sure it reads from the updates channel in the `_run()` method before doing any other work, otherwise it might be running with a missing configuration until the update message arrives. I think we should put this more clearly in the docs, including the example. But we can do it as a followup PR.

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [reply src/frequenz/dispatch/_actor_runner.py]: After talking to @daniel-zullo-frequenz we agreed that in the future his use case should use only one central actor to handle dispatches that spawns more actors as needed, but that would mean a big refactoring we can start right now, so we would leave the support for multiple actors here for now. What we can do is accept `actor: Actor | Set[Actor]`, so for all normal use cases people can pass just one actor, and document the use of multiple actors is discouraged and will be removed in the future.

Comment (llucax) [reply src/frequenz/dispatch/_actor_runner.py]: Oh, I see you removed the support for multiple actors already. I didn't meant we should remove it, but we should *figure out if we really need to ...* :(

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [reply src/frequenz/dispatch/_actor_runner.py]: Yeah, that could work too, I think we discussed `options` at some point but wasn't perfectly accurate as some "options" might not be *optional*. I guess `inputs` is an improvement in that sense. Maybe `parameters` could work too.

File: src/frequenz/dispatch/_managing_actor.py
Comment (llucax) [inline src/frequenz/dispatch/_managing_actor.py]:
```suggestion
        # Start actor and supporting actor, give each an updates channel receiver
```

File: tests/test_mananging_actor.py
Comment (llucax) [inline tests/test_mananging_actor.py]:
```suggestion
    updates_receiver: Receiver[DispatchUpdate]
```

File: tests/test_mananging_actor.py
Comment (llucax) [inline tests/test_mananging_actor.py]:
```suggestion
    config_channel = Broadcast[DispatchUpdate](name="dispatch updates test channel")
```

File: tests/test_mananging_actor.py
Comment (llucax) [inline tests/test_mananging_actor.py]:
```suggestion
        updates_sender=updates_channel.new_sender(),
```

File: tests/test_mananging_actor.py
Comment (llucax) [inline tests/test_mananging_actor.py]:
```suggestion
        updates_receiver=updates_channel.new_receiver(),
```

File: tests/test_mananging_actor.py
Comment (llucax) [inline tests/test_mananging_actor.py]:
```suggestion
    event = await test_env.updates_receiver.receive()
```

File: tests/test_mananging_actor.py
Comment (llucax) [inline tests/test_mananging_actor.py]:
```suggestion
    event = await test_env.updates_receiver.receive()
```

**Review by llucax — COMMENTED** [review]
A few more instances of *config* and we need support for multiple actors for now (temporarily). I will also create issues for the stuff we can address in the future.

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [reply src/frequenz/dispatch/_actor_runner.py]: * #55

File: src/frequenz/dispatch/_actor_runner.py
Comment (llucax) [reply src/frequenz/dispatch/_actor_runner.py]: * #56

File: src/frequenz/dispatch/_managing_actor.py
Comment (llucax) [reply src/frequenz/dispatch/_managing_actor.py]: * https://github.com/frequenz-floss/frequenz-dispatch-python/issues/56

**Comment by Marenz** [comment]
* Added multiple actors back
* fixed confi -> up stuff

File: src/frequenz/dispatch/_dispatch.py
Comment (llucax) [reply src/frequenz/dispatch/_dispatch.py:195]: * https://github.com/frequenz-floss/frequenz-client-dispatch-python/issues/93

File: src/frequenz/dispatch/actor.py
Comment (llucax) [reply src/frequenz/dispatch/actor.py:112]: * #57

File: src/frequenz/dispatch/actor.py
Comment (llucax) [inline src/frequenz/dispatch/actor.py:64]: * #58

**Review by llucax — DISMISSED** [review]

**Review by llucax — DISMISSED** [review]
So this should be ready for merge. @daniel-zullo-frequenz wants to do a release soon using this, so ideally we should have a release too, but I guess for that we first need to release some dependencies too, right?

We need to merge changes in the `client-dispatch` and for that some changes in the `client-base`, right? We had some changes in `channels` too but those were already released in v1.2.0, correct?

**Comment by Marenz** [comment]
> [..] we first need to release some dependencies too, right?

Yes.

> We need to merge changes in the client-dispatch and for that some changes in the client-base, right? 

Almost. The changes we merged now actually recreate the streamer object so we don't need any base-client changes.
I have already released the `client-dispatch` version we need now.

> We had some changes in channels too but those were already released in v1.2.0, correct?

Correct. Those are also reflected in the current pyproject.py.


I pushed the last dependency changes and a little new feature so that the actor properly re-reads the dispatches upon disconnect now.

**Review by llucax — DISMISSED** [review]

File: pyproject.toml
Comment (llucax) [inline pyproject.toml:168]: Maybe we should remove this and keep the default as it is?

File: tests/test_mananging_actor.py
Comment (llucax) [inline tests/test_mananging_actor.py]: Also if it fails even with session, I would leave it to the default too here.

File: tests/test_frequenz_dispatch.py
Comment (llucax) [inline tests/test_frequenz_dispatch.py]: Also if it fails even with session, I would leave it to the default too here.

File: pyproject.toml
Comment (Marenz) [reply pyproject.toml:168]: Then we get a depreciation warning saying the behavior will change and one should set a default

Comment (llucax) [reply pyproject.toml:168]: Oh, OK, I'm confused now. So this has nothing to do with trying to make tests pass and is just a new deprecation message in a new version of `pytest-asyncio`? If so it would be probably nice to put it in a separate commit with a corresponding commit message.

Comment (Marenz) [reply pyproject.toml:168]: I def. tried this to make the test pass, bit it also fixes a deprecation warning, it just didn't help though.

Comment (llucax) [reply pyproject.toml:168]: OK, this seems to be related to an update to https://github.com/pytest-dev/pytest-asyncio/releases/tag/v0.24.0. Issue: https://github.com/pytest-dev/pytest-asyncio/issues/924

**Review by llucax — DISMISSED** [review]

**Review by llucax — APPROVED** [review]

**Comment by llucax** [comment]
It looks like now cross-arch tests are consistently failing. This must be definitely a timing issue, the slowest the system is, the more likely we hit this bug and tests hang. Created an issue about it:

* https://github.com/frequenz-floss/frequenz-dispatch-python/issues/66

**Pull request merged**
